### PR TITLE
Constraint tool

### PR DESCRIPTION
**Pull request by Flova**
Add tool to define time constraints and get warned if they are unsatisfied.

### CODE REVIEW

File: bitbots_time_constraint/src/bitbots_time_constraint/check_constraints.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+        self.__subscriber = None        # type: rospy.Subscriber
+        self.__subscribe()
+
+    def __subscribe(self, backoff=1.0):
+        """
+        Try to subscribe to the set topic
+        :param backoff: How long to wait until another try
+        """
+        data_class, _, _ = rostopic.get_topic_class(self.topic)
```
Comment (NFiedler) [inline bitbots_time_constraint/src/bitbots_time_constraint/check_constraints.py]: i would do this with [get_published_topics](https://docs.ros.org/api/rospy/html/rospy-module.html#get_published_topics).

File: bitbots_time_constraint/src/bitbots_time_constraint/check_constraints.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            # topic is known
+            self.__subscriber = rospy.Subscriber(self.topic, data_class, self.__message_callback, queue_size=1, tcp_nodelay=True)
+            rospy.loginfo('Subscribed to topic {}'.format(self.topic))
+
+        else:
+            # topic is not yet known
+            rospy.loginfo('Topic {} is not yet known. Retrying in {} seconds'.format(self.topic, int(backoff)))
+            rospy.Timer(
+                rospy.Duration(int(backoff)),
+                lambda event: self.__subscribe(backoff * 1.2),
```
Comment (NFiedler) [inline bitbots_time_constraint/src/bitbots_time_constraint/check_constraints.py]: why ```* 1.2```?

File: bitbots_time_constraint/src/bitbots_time_constraint/check_constraints.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        rospy.init_node("bitbots_time_constraint", anonymous=False)
+
+        config_path = os.path.join(self.package_path, "config/constraints.yaml")
+
+        with open(config_path, 'r') as stream:
+            try:
+                constraint_config = yaml.safe_load(stream)
+            except yaml.YAMLError as exc:
+                rospy.roserr("Config not found!!!!")
+                rospy.signal_shutdown("")
```
Comment (NFiedler) [inline bitbots_time_constraint/src/bitbots_time_constraint/check_constraints.py]: Why? The ROS parameter server is able to read YAML files and you are using a launch file. Please use the ROS parameter server.

File: bitbots_time_constraint/src/bitbots_time_constraint/check_constraints.py (hunk 1 above)
Comment (Flova) [reply bitbots_time_constraint/src/bitbots_time_constraint/check_constraints.py]: That is not what I want.

File: bitbots_time_constraint/src/bitbots_time_constraint/check_constraints.py (hunk 2 above)
Comment (Flova) [reply bitbots_time_constraint/src/bitbots_time_constraint/check_constraints.py]: Its recursively called and with each call, the retry time increases with this magic value.

File: bitbots_time_constraint/package.xml [hunk 4]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0"?>
+<package format="2">
+  <name>bitbots_time_constraint</name>
+  <version>1.0.0</version>
+  <description>TODO</description>
```
Comment (timonegk) [inline bitbots_time_constraint/package.xml]: Please add a description

File: bitbots_time_constraint/package.xml [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  <build_depend>message_generation</build_depend>
+
+  <depend>rospy</depend>
+
+  <exec_depend>std_msgs</exec_depend>
+
+  <export>
+    <bitbots_documentation>
+      <language>python3</language>
+      <status>untested</status>
```
Comment (timonegk) [inline bitbots_time_constraint/package.xml]: Please update status

File: bitbots_time_constraint/src/bitbots_time_constraint/check_constraints.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    def __message_callback(self, data):
+        self.latest_timestamp = rospy.Time.now()
+
+
+class ConstraintChecker:
+    def __init__(self):
+        rospack = rospkg.RosPack()
+        self.package_path = rospack.get_path('bitbots_time_constraint')
+
+        rospy.init_node("bitbots_time_constraint", anonymous=False)
```
Comment (timonegk) [inline bitbots_time_constraint/src/bitbots_time_constraint/check_constraints.py]: You want to set anonymous to True to be able to start multiple instances of this node (one on each host)

**Pull request merged**